crates/matrix-sdk-base/src/rooms/normal.rs

```diff
@@ -222,6 +226,18 @@ impl From<&MembershipState> for RoomState {
 /// try to behave similarly here.
 const NUM_HEROES: usize = 5;
 
+/// A filter to remove our own user and the users specified in the member hints
+/// state event, so called service members, from the list of heroes.
+///
+/// The heroes will then be used to calculate a display name for the room if one
+/// wasn't explicitly defined.
+fn heroes_filter<'a>(
+    own_user_id: &'a UserId,
+    member_hints: &'a MemberHintsEventContent,
+) -> impl Fn(&UserId) -> bool + use<'a> {
+    move |user_id| user_id != own_user_id && !member_hints.service_members.contains(user_id)
+}
```

Comment:
This is just a standard partially applied function, nothing really fancy going on.

But we're not interested in the members in both cases, in the case where we do have a summary we're only fetching individual members for which we didn't receive the name of the hero as part of the summary:

https://github.com/matrix-org/matrix-rust-sdk/blob/0ecbbc2e6e1bce101dc1b59efffd6a0276552362/crates/matrix-sdk-base/src/rooms/normal.rs#L706-L722

Reply:
Ohh you're right, members are not loaded in this case.

(It might also mean that if you have a room with 10 members, and the summary retrieved from the server contains only 6 but it includes the service member, then the final room name will include 5 heroes, while it could include another non-service-member? Maybe a feedback for the MSC, unless the server knows about service members.)

Could we have a freestanding function iterating over user_ids and returning an `impl Iterator<Item = &UserId>` that contains the filtered ids, maybe?

Reply:
> Why?

A detour for programmers, which is cognitive overhead; more complicated type checking for the compiler; I do prefer simplicity in the form of having a single line repeated in two places, rather than a new function returning a common filter — sometimes, the good intent of sharing code ends up confusing the reader a bit too much. But it's alright, as you seem to have a strong opinion on this, and this ain't the hill I want to die on :+1:

Reply:
>But it's alright, as you seem to have a strong opinion on this, and this ain't the hill I want to die on 👍

I mean you didn't suggest any alternative, the first thing you suggested didn't work, the second suggestions I'm fishing for meaning but I don't really know what you meant.

> I do prefer simplicity in the form of having a single line repeated in two places

Ok now you do suggest something, and I have to disagree here. I almost missed that the filtering needs to happen in two places, if we do this and somebody else has to modify the filter they'll very likely miss this fact.